sound/soc/sof/ops.c

```diff
@@ -161,3 +161,92 @@ void snd_sof_dsp_panic(struct snd_sof_dev *sdev, u32 offset)
 	snd_sof_trace_notify_for_error(sdev);
 }
 EXPORT_SYMBOL(snd_sof_dsp_panic);
+
+/* This is for getting ref count for a DSP core and power on it if needed */
+int snd_sof_dsp_core_get(struct snd_sof_dev *sdev, u32 core_idx)
+{
+	u32 target_cores_mask = sdev->enabled_cores_mask | BIT(core_idx);
+	int ret = 0;
+
+	mutex_lock(&sdev->cores_status_mutex);
+
+	/* return if already powered on */
+	if (sdev->core_refs[core_idx] > 0) {
+		dev_vdbg(sdev->dev, "core_get: enabled_cores_mask 0x%x, core_refs[%d] %d, no need to power up\n",
+			 sdev->enabled_cores_mask, core_idx, sdev->core_refs[core_idx]);
+		sdev->core_refs[core_idx]++;
+		goto done;
+	}
+
+	dev_vdbg(sdev->dev, "core_get: enabled_cores_mask 0x%x, core_refs[%d] %d, powering it up...\n",
+		 sdev->enabled_cores_mask, core_idx, sdev->core_refs[core_idx]);
+	/* power up the core that this pipeline is scheduled on */
+	ret = snd_sof_dsp_core_power_up(sdev, BIT(core_idx));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: powering up pipeline schedule core %d\n",
+			core_idx);
+		goto done;
+	}
+
+	/* Now notify DSP that the core power status changed */
+	ret = snd_sof_ipc_core_enable(sdev, target_cores_mask);
+	if (ret < 0) {
+		/* power down the core to reflect the status */
+		snd_sof_dsp_core_power_down(sdev, BIT(core_idx));
+		goto done;
+	}
+
+	/* update core ref count and enabled_cores_mask */
+	sdev->core_refs[core_idx]++;
+	sdev->enabled_cores_mask = target_cores_mask;
```

Comment:
again do we really need both an enabled bit and a refcount. this seems like duplication of information.